### PR TITLE
fix: split http instrumentation to core logic and http lib integration

### PR DESCRIPTION
**Pull request by sagivoululumigo**
This is the first stage of adding native node fetch instrumentation.

This PR splits the existing http library instrumentation to the core logic of tracing a generic http request (the `BaseHttp` class) and the instrumentation connecting that logic to the `http` / `https` libraries (the `Http` class).

The next stage will be in a separate PR that will use the `BaseHttp` class to integrate to a new http library called `undici` (This is the library that native node fetch uses behind the scenes)

### CODE REVIEW

File: src/hooks/baseHttp.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+  extractBodyFromEmitSocketEvent,
+  extractBodyFromWriteOrEndFunc,
+  httpDataToString,
+} from './httpUtils';
+import { URL } from 'url';
+import { parse as parseQuery } from 'querystring';
+import { shallowMask } from '../utils/payloadStringify';
+import { Agent } from './http';
+
+export const hostBlaclist = new Set(['127.0.0.1']);
```
Comment (harelmo-lumigo) [inline src/hooks/baseHttp.ts:39]:
```suggestion
export const hostBlacklist = new Set(['127.0.0.1']);
```

File: src/hooks/baseHttp.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+  path?: string;
+  truncated?: boolean;
+  port?: number;
+  uri?: string;
+  method?: string;
+  protocol?: string;
+  sendTime?: number;
+};
+
+export class BaseHttp {
```
Comment (harelmo-lumigo) [inline src/hooks/baseHttp.ts:87]: Since all these methods here are static, what's the value of having it inside a class? it can be broken into standalone functions exported from this module IMO

File: src/hooks/baseHttp.ts [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+    const uri = `${host}${path}${search}`;
+
+    return {
+      truncated: false,
+      path,
+      port,
+      uri,
+      host,
+      body: '', // XXX Filled by the httpRequestEndWrapper or httpRequestEmitBeforeHookWrapper ( / Write)
```
Comment (harelmo-lumigo) [inline src/hooks/baseHttp.ts:370]: Is XXX placeholder for something?
```suggestion
      body: '', // XXX Filled by the httpRequestEndWrapper or httpRequestEmitBeforeHookWrapper ( / Write)
```

Comment (harelmo-lumigo) [inline src/hooks/baseHttp.ts:370]: Placeholder?
```suggestion
      body: '', // Filled by the httpRequestEndWrapper or httpRequestEmitBeforeHookWrapper ( / Write)
```

**Review by harelmo-lumigo — APPROVED** [review]

File: src/hooks/baseHttp.test.js [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+  test('isBlacklisted', () => {
+    const host = 'asdf';
+    const edgeHost = 'us-east-x.lumigo-tracer-edge.golumigo.com';
+    TracerGlobals.setTracerInputs({ ...TracerGlobals.getTracerInputs(), edgeHost });
+    expect(BaseHttp.isBlacklisted(host)).toBe(false);
+    expect(BaseHttp.isBlacklisted(edgeHost)).toBe(true);
+  });
+
+  test('aggregateRequestBodyToSpan ->  happy flow', () => {
```
[the remark this answers is not in the document]
Comment (sagivoululumigo) [reply src/hooks/baseHttp.test.js:30]: This test is copied from the `http.test.js` file

File: src/hooks/baseHttp.ts (hunk 1 above)
Comment (sagivoululumigo) [reply src/hooks/baseHttp.ts:39]: Done

File: src/utils/w3cUtils.test.ts [hunk 5]
```diff
@@ -58,6 +58,13 @@ describe('Utf8Utils', () => {
     expect(headers[TRACESTATE_HEADER_NAME]).toEqual(`lumigo=${parts[2]}`);
   });
 
+  test('addW3CTracePropagator -> no new headers added', () => {
```
[the remark this answers is not in the document]
Comment (sagivoululumigo) [reply src/utils/w3cUtils.test.ts:61]: Done

File: src/utils/w3cUtils.ts [hunk 6]
```diff
@@ -17,15 +17,23 @@ export const shouldSkipTracePropagation = (headers: Record<string, string>): boo
...
-  return headers;
+  return {
+    [TRACEPARENT_HEADER_NAME]: getTraceId(headers, getCurrentTransactionId(), messageId),
+    [TRACESTATE_HEADER_NAME]: getTraceState(headers, messageId),
+  };
+};
+
+export const addW3CTracePropagator = (headers: Record<string, string>): Record<string, string> => {
+  const additionalHeaders = getW3CTracerPropagatorAdditionalHeaders(headers);
+  return Object.assign(headers, additionalHeaders);
```
[the remark this answers is not in the document]
Comment (sagivoululumigo) [reply src/utils/w3cUtils.ts:36]: Yes I want to edit the headers in place, I don't want to change the previous behavior

File: src/hooks/baseHttp.ts (hunk 3 above)
Comment (sagivoululumigo) [reply src/hooks/baseHttp.ts:370]: IDK this was directly copied from the previous implementation...
Anyways removed the XXX and kept the rest of the comment

File: src/spans/awsSpan.ts [hunk 7]
```diff
@@ -416,7 +417,7 @@ export const getHttpSpan = (
     HTTP_SPAN
   );
 
-  let info = Object.assign({}, basicHttpSpan.info, {
+  const info = Object.assign({}, basicHttpSpan.info, {
     httpInfo,
     ...awsServiceData,
   });
```
[the remark this answers is not in the document]
Comment (sagivoululumigo) [reply src/spans/awsSpan.ts:423]: Note that we are assigning `Key: "httpInfo", Value: httpInfo` to the final object, and not the fields in `httpInfo` directly. So fixed the line to be:
```
const info = Object.assign({}, basicHttpSpan.info, { httpInfo }, awsServiceData);
```

**Review by GuyMoses — APPROVED** [review]
since you’ve moved a lot of code, it's hard to follow the PR.
I suggest to mark areas that were only relocated, so that the the CR could be focused on altered/ new lines mainly

File: src/hooks/baseHttp.ts [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+    const requestData = BaseHttp.parseHttpRequestOptions(options, url);
+    returnData.requestData = requestData;
+
+    if (isTimeoutTimerEnabled()) {
+      const httpSpan = getHttpSpan(transactionId, awsRequestId, requestRandomId, requestData);
+      SpansContainer.addSpan(httpSpan);
+      returnData.httpSpan = httpSpan;
+    }
+    return returnData;
```
Comment (sagivoululumigo) [inline src/hooks/baseHttp.ts:151]: Critical code segment!
This is in essence the same code as the start hook in the http lib instrumentation, just made it general to match any http library.

File: src/hooks/baseHttp.ts [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+        const body = extractBodyFromEmitSocketEvent(socketEventArgs);
+        BaseHttp.aggregateRequestBodyToSpan(
+          body,
+          requestData,
+          currentSpan,
+          getEventEntitySize(true)
+        );
+      }
+      GlobalDurationTimer.stop();
+    };
```
Comment (sagivoululumigo) [inline src/hooks/baseHttp.ts:231]: Critical code segment!
Basically copied from the old http instrumentation

File: src/hooks/baseHttp.ts [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+          Object.assign({ truncated }, responseData)
+        );
+        if (httpSpan.id !== requestRandomId) {
+          // In Http case, one of our parser decide to change the spanId for async connection
+          SpansContainer.changeSpanId(requestRandomId, httpSpan.id);
+        }
+        SpansContainer.addSpan(httpSpan);
+      }
+      GlobalDurationTimer.stop();
+    };
```
Comment (sagivoululumigo) [inline src/hooks/baseHttp.ts:300]: Copied almost 100% from the old implementation, small tweaks
